fdmi/source_dock_fom.c

```diff
@@ -345,20 +346,42 @@ M0_INTERNAL void m0_fdmi__src_dock_fom_wakeup(struct fdmi_sd_fom *sd_fom)
 	M0_LEAVE();
 }
 
+static void fdmi__src_dock_timer_fom_wakeup(struct fdmi_sd_timer_fom *timer_fom)
+{
+	struct m0_fom *fom;
+
+	M0_ENTRY("sd_timer_fom %p", timer_fom);
+	M0_PRE(timer_fom != NULL);
+
+	fom = &timer_fom->fstf_fom;
+
+	if (timer_fom->fstf_wakeup_ast.sa_next == NULL) {
+		timer_fom->fstf_wakeup_ast = (struct m0_sm_ast) {
+			.sa_cb    = wakeup_iff_waiting,
+			.sa_datum = fom
+		};
+		m0_sm_ast_post(&fom->fo_loc->fl_group,
```

Comment:
Can we ensure if this ast is not posted FDMI service is stopped already ?

Reply:
We have a semaphore to wait on.
```
	/* Wait for timer fom finished */
	m0_semaphore_down(&src_dock->fsdc_sd_timer_fom.fstf_shutdown);
	m0_semaphore_fini(&src_dock->fsdc_sd_timer_fom.fstf_shutdown);
```

Only when that semaphore is UP, the FDMI service will stop.
Before the semaphore is UP, it is OK to post AST to an active FOM.
The same for source dock fom.